### PR TITLE
Implementation of M190 bed temp hysteresis

### PR DESCRIPTION
**Pull request by scalez**
Add a residency timer for the bed in `M190`, just as there is for extruders in `M109`.

### CODE REVIEW

File: Marlin/Configuration.h [hunk 1]
```diff
@@ -192,11 +192,15 @@
 //#define TEMP_SENSOR_1_AS_REDUNDANT
 #define MAX_REDUNDANT_TEMP_SENSOR_DIFF 10
 
-// Actual temperature must be close to target for this long before M109 returns success
+// Actual temperature must be close to target for this long before M109/M190 returns success
 #define TEMP_RESIDENCY_TIME 10  // (seconds)
 #define TEMP_HYSTERESIS 3       // (degC) range of +/- temperatures considered "close" to the target one
 #define TEMP_WINDOW     1       // (degC) Window around target to start the residency timer x degC early.
```
Comment (thinkyhead) [inline Marlin/Configuration.h]: Might as well do this for bed leveling too…
```cpp
// Actual temperature must be close to target for this long before M104/M140 returns success
```

**Comment by thinkyhead** [comment]
Looks great! All it needs is to add the same lines to all `Configuration.h` files that were added to the main one.

**Comment by Blue-Marlin** [comment]
Since when M104/140 should wait for anything? (https://github.com/MarlinFirmware/Marlin/pull/3477#discussion_r59480738)

**Comment by jbrazio** [comment]
M104 ? He is patching M109 here.

**Comment by scalez** [comment]
@jbrazio: @Blue-Marlin is responding to @thinkyhead's [#3477](https://github.com/MarlinFirmware/Marlin/pull/3477#discussion_r59480738) comment. Not sure what @thinkyhead means there, especially about the bed leveling, maybe he had a few too many last night haha.

@thinkyhead could you pls elaborate? Also, what do you mean by it needing to be added "_to the main ones_" do you mean the example configuration headers?

**Comment by thinkyhead** [comment]
> especially about the bed leveling

Obviously he meant "bed heating" but was in some kind of a hurry. Likewise the confabulation of `M104/M140` with `M109/M190` – which is a silly numbering scheme anyway.

**Comment by thinkyhead** [comment]
> what do you mean…?

As I said… "All it (this PR) needs is (for you) to **add the same lines** to all `Configuration.h` files (in the `example_configurations` folder) **that were added to the main one** (i.e., `Marlin/Configuration.h`).

Whenever a change is made to `Marlin/Configuration.h` we must likewise change all other `Configuration.h` files that are included with Marlin so they are all synchronized. The new option may be enabled in some configurations, if it makes sense.

File: Marlin/Configuration.h [hunk 2]
```diff
@@ -192,11 +192,15 @@
 //#define TEMP_SENSOR_1_AS_REDUNDANT
 #define MAX_REDUNDANT_TEMP_SENSOR_DIFF 10
 
-// Actual temperature must be close to target for this long before M109 returns success
+// Actual temperature must be close to target for this long before M109/M190 returns success
```
Comment (thinkyhead) [inline Marlin/Configuration.h]: Sorry, let me try again. Change this to refer especially to the extruder…

``` cpp
// Extruder temperature must be close to target for this long before M109 returns success
```

File: Marlin/Configuration.h [hunk 3]
```diff
@@ -192,11 +192,15 @@
...
 #define TEMP_RESIDENCY_TIME 10  // (seconds)
 #define TEMP_HYSTERESIS 3       // (degC) range of +/- temperatures considered "close" to the target one
 #define TEMP_WINDOW     1       // (degC) Window around target to start the residency timer x degC early.
 
+#define TEMP_BED_RESIDENCY_TIME 10  // (seconds)
```
Comment (thinkyhead) [inline Marlin/Configuration.h]: But then add a comment specific to the bed here, for clarity…
```cpp
// Bed temperature must be close to target for this long before M190 returns success
```

File: Marlin/Configuration.h (hunk 1 above)
Comment (thinkyhead) [inline Marlin/Configuration.h:199]: But then add a comment specific to the bed here, for clarity…

``` cpp
// Bed temperature must be close to target for this long before M190 returns success
```

**Comment by scalez** [comment]
@thinkyhead thanks for clearing that up!

I'll place it in the example config headers but I'm not sure which machines should have it enabled. We're using it on our TAZ 6 and Mini but we're not quite ready to make a PR for their configs yet. Unless you know off hand which ones should have it enabled and what values they should use, I can just stick them in and leave `TEMP_BED_RESIDENCY_TIME` commented/undefined.

**Comment by jbrazio** [comment]
I would leave it off by default by setting the time to zero as this introduces new behavior and we're on a RC cycle.

**Comment by scalez** [comment]
@thinkyhead @jbrazio I pushed a commit with the additions to the example config headers.

I set `TEMP_BED_RESIDENCY_TIME` to zero in each of the example configs, but keep in mind that this will still compile the code inside the `#ifdef TEMP_BED_RESIDENCY_TIME` conditionals, only commenting the line which defines `TEMP_BED_RESIDENCY_TIME` in Configuration.h will disable it (won't be compiled).

**Comment by jbrazio** [comment]
@scalez Could you try with `#if ENABLED(TEMP_BED_RESIDENCY_TIME)` ?

**Comment by scalez** [comment]
@jbrazio Yes I could, although `TEMP_RESIDENCY_TIME` in M109 is using `#ifdef`. Imo the two should be consistent between one another. Both could be changed to use 'ENABLED', should that be part of this PR?

**Comment by jbrazio** [comment]
If the usage of `ENABLED()` solves the issue, then I would suggest to update both for consistency.

**Comment by thinkyhead** [comment]
@scalez @jbrazio No, the `ENABLED` macro isn't appropriate here, because `TEMP_RESIDENCY_TIME` and `TEMP_BED_RESIDENCY_TIME` are assigned numerical values. The `ENABLED` macro only applies to switches that are `#defined` or `#undef`, `true` or `false`, and `1` or `0`.

However, @scalez is onto something here, which I also just noticed. All the lines that read…

``` cpp
#ifdef TEMP_RESIDENCY_TIME
#ifdef TEMP_BED_RESIDENCY_TIME
```

…should be changed to…

``` cpp
#if defined(TEMP_RESIDENCY_TIME) && TEMP_RESIDENCY_TIME > 0
#if defined(TEMP_BED_RESIDENCY_TIME) && TEMP_BED_RESIDENCY_TIME > 0
```

…or because of the way pre-compiler macros work, they could also just be…

``` cpp
#if TEMP_RESIDENCY_TIME > 0
#if TEMP_BED_RESIDENCY_TIME > 0
```

…because undefined symbols in `#if` directives evaluate to zero.

**Comment by thinkyhead** [comment]
I suppose the `ENABLED` / `DISABLED` macros could be expanded to handle cases like this:
```cpp
#define ENABLED(b) _CAT(SWITCH_ENABLED_, b) || (b > 0)
#define DISABLED(b) (!_CAT(SWITCH_ENABLED_, b)) || (b == 0)
```

**Comment by jbrazio** [comment]
@thinkyhead thanks for the clarification, let's go for the simplier `#if TEMP_RESIDENCY_TIME > 0` which IMO is readable enough.

**Comment by scalez** [comment]
@thinkyhead @jbrazio let me know what you think of https://github.com/MarlinFirmware/Marlin/pull/3477/commits/8ec8d59df8ad647da81f3edb39866cf508ef4560.

Also, I just added the watch temp protection feature to the bed, check this commit: https://github.com/alephobjects/Marlin/commit/d957f9543734e339aaf27114f7b26b00814e0823

Let me know if you want me to make a PR for it as well.

**Comment by jbrazio** [comment]
I really like the idea to have a consistent thermal code, wither we're using a heater, a hot bed, a hot box etc because in the long run it will make easier for us to build the `Heater`, `Cooler` (FAN etc) object to interact with some kind of `ThermalManager`.

So d957f95 is another step forward.

**Comment by thinkyhead** [comment]
@scalez Good work – ready to go. And the temp watch for the bed looks good too. I look forward to the PR for that also.
